src/manifolds/Segre.jl

```diff
@@ -0,0 +1,473 @@
+````
+where ``\nu``, ``\xi \in T_{\lambda} ℝ^{+} = ℝ`` and ``u_i``, ``v_i \in T_{x_i} S^{n_i - 1} \subset ℝ^{n_i}``.
+"""
+function inner(M::Segre{ℝ,V}, p, u, v) where {V}
+    return u[1][1] * v[1][1] + p[1][1]^2 * dot(u[2:end], v[2:end])
+end
+
+@doc raw"""
+    function norm(M::Segre{𝔽, V}, p, v)
+
+Norm of tangent vector ``v`` at ``p``.
+"""
+function norm(M::Segre{𝔽,V}, p, v) where {𝔽,V}
+    return sqrt(inner(M, p, v, v))
+end
+
+"""
+    function rand(M::Segre{ℝ, V}; vector_at=nothing)
+"""
+function rand(M::Segre{ℝ,V}; vector_at=nothing) where {V}
+    if isnothing(vector_at)
+        lambda = abs.(rand(Euclidean(1)))
+        xs = [rand(Sphere(n - 1)) for n in V]
+        return [lambda, xs...]
+    else
+        @assert(is_point(M, vector_at))
+
+        lambdadot = rand(Euclidean(1); vector_at=vector_at[1])
+        xdots = [rand(Sphere(n - 1); vector_at=vector_at[i + 1]) for (i, n) in enumerate(V)]
+        return [lambdadot, xdots...]
+    end
+end
+
+@doc raw"""
+    function embed(M::Segre{𝔽, V}, v)
+
+Embed ``p \doteq (\lambda, x_1, \dots, x_d)`` in ``𝔽^{n_1 \times \dots \times n_d}`` using the Krönecker product:
+````math
+    (\lambda, x_1, \dots, x_d) \mapsto \lambda x_1 \otimes \dots \otimes x_d.
+````
+"""
+function embed(M::Segre{𝔽,V}, p) where {𝔽,V}
+    return kronecker(p...)[:]
+end
+
+@doc raw"""
+    function embed_vector(M::Segre{𝔽, V}, p, v)
+
+Embed tangent vector ``v = (\nu, u_1, \dots, u_d)`` at ``p \doteq (\lambda, x_1, \dots, x_d)`` in ``𝔽^{n_1 \times \dots \times n_d}`` using the Krönecker product:
+````math
+    (\nu, u_1, \dots, u_d) \mapsto \nu x_1 \otimes \dots \otimes x_d + \lambda u_1 \otimes x_2 \otimes \dots \otimes x_d + \dots + \lambda x_1 \otimes \dots \otimes x_{d - 1} \otimes u_d.
+````
+"""
+function embed_vector(M::Segre{𝔽,V}, p, v) where {𝔽,V}
+
+    # Product rule
+    return sum([
+        kronecker([i == j ? xdot : x for (j, (x, xdot)) in enumerate(zip(p, v))]...)[:] for
+        (i, _) in enumerate(p)
+    ])
+end
+
+@doc raw"""
+    function m(M::Segre{ℝ, V}, p, q)
+
+Let ``p \doteq (\lambda, x_1, \dots, x_d)``, ``q \doteq (\mu, y_1, \dots, y_d) \in \mathcal{S}``.
+Then
+````math
+    m(p, q) = \sqrt{\sphericalangle(x_1, y_1)^2 + \dots + \sphericalangle(x_d, y_d)^2}
+````
+is the distance between ``(x_1, \dots, x_d)`` and ``(y_1, \dots, y_d)`` on ``S^{n_1 - 1} \times \dots \times S^{n_d - 1}``.
+"""
+function m(M::Segre{ℝ,V}, p, q) where {V}
+    return sqrt(
+        sum([distance(Sphere(n - 1), x, y)^2 for (n, x, y) in zip(V, p[2:end], q[2:end])]),
+    )
+end
+
+"""
+    function connected_by_geodesic(M::Segre{ℝ, V}, p, q)
+
+Check if two points, `p` and `q`, can be connected by a geodesic.
+"""
+function connected_by_geodesic(M::Segre{ℝ,V}, p, q) where {V}
+    q_ = closest_representation(M, p, q)
+
+    return m(M, p, q_) < pi
+end
+
+"""
+    function closest_representation(M::Segre{ℝ, V}, p, q)
+
+Find the representation of `q` that is closest to `p`.
```

Comment:
Please explain this, I am not sure what `p` and `q` are – not what the closest representation here refers to.
Is q not on M?